src/main/scala/components/csr.scala

```diff
@@ -0,0 +1,506 @@
+/* Describes register file that maintains machine state */
+
+package dinocpu
+
+import chisel3._
+import collection.mutable.LinkedHashMap
+import chisel3.util._
+import chisel3.util.BitPat
+import chisel3.util.experimental.BoringUtils
+import Util._
+
+import scala.math._
+
+object MCauses {
+  //with interrupt
+  val machine_soft_int = 0x80000003
+  val machine_timer_int = 0x80000007
+  val machine_ext_int = 0x8000000b
+
+  //non interrupt
+  val misaligned_fetch = 0x0
+  val fetch_access = 0x1
+  val illegal_instruction = 0x2
+  val breakpoint = 0x3
+  val misaligned_load = 0x4
+  val load_access = 0x5
+  val misaligned_store = 0x6
+  val store_access = 0x7
+  val machine_ecall = 0xb
+  val all = {
+    val res = collection.mutable.ArrayBuffer[Int]()
+    res += machine_soft_int
+    res += machine_timer_int
+    res += machine_ext_int
+
+    res += misaligned_fetch
+    res += fetch_access
+    res += illegal_instruction
+    res += breakpoint
+    res += misaligned_load
+    res += load_access
+    res += misaligned_store
+    res += store_access
+    res += machine_ecall
+    res.toArray
+  }
+}
+
+object MCSRs {
+  //machine information registers
+  val mvendorid = 0xf11 //vendor id
+  val marchid = 0xf12 //architecture id
+  val mimpid = 0xf13 //implementation id
+  val mhartid = 0xf14 //hardware thread id
+  //machine trap setup
+  val mstatus = 0x300 //machine status reg
+  val misa = 0x301 //isa and extensions
+  val medeleg = 0x302 //machine exception delegation reg
+  val mideleg = 0x303 //machine interrupt delegation reg
+  val mie = 0x304 //machine iterrupt-enable reg
+  val mtvec = 0x305 //machine trap handler base address
+  val mcounteren = 0x306 //machine counter enable
+
+  //machine trap handling
+  val mscratch = 0x340 //scratch reg for machine trap handlers
+  val mepc = 0x341 //machine exception program counter
+  val mcause = 0x342 //machine trap cause
+  val mtval = 0x343 //machine bad address or instruction
+  val mip = 0x344 //machine interrupt pending
+
+  //machine memory protection
+  //DONT NEED
+  
+  //machine counter/timers
+  val mcycle = 0xb00 //machine cycle counter
```

Comment:
What are these numbers?!?!

Reply:
index into the register file for that particular register